# SnapATAC from Cell Ranger ATAC

This pipeline is usefull if you already have the outputs generated by running Cell Ranger ATAC.

The pipeline takes Cell Ranger ATAC's position-sorted BAM file and runs the following:

1. Barcode attachment
1. Snap preprocessing
1. Add cell-by-bin matrix

The final outputs includes:

1. Snap file
2. Sorted BAM file

You can take the output snap file and run the secondary analysis pipeline.

## Requirements

- `cromwell-tools`
- Service account key
- For WDL validation, the bash scripts expect `womtool-40.jar` to be in `~/Applications/` (fixme)

## How to Run Examples

The example is 10x fresh cortex from adult mouse brain (P50):
https://support.10xgenomics.com/single-cell-atac/datasets/1.1.0/atac_v1_adult_brain_fresh_5k

Below will submit a job to CaaS:

```bash
$ ./submit.sh \
    -k ~/pipelines_surge_credentials.json
```

## WDL Validation

```bash
$ ./validate.sh
```
